compiler/src/lnntoamm/Statement.ts

```diff
@@ -0,0 +1,785 @@
+    let dec = stmts.pop();
+    if (dec instanceof Dec) {
+      metadata.variables[name] = metadata.var(dec.name);
+      metadata.variables[dec.name] = null;
+      dec.ty.constrain(ty);
+      dec.mutable = mutable;
+      dec.name = name;
+    } else if (dec instanceof Ref) {
+      const upTy = dec.ty;
+      dec = new Dec(
+        ast,
+        mutable,
+        name,
+        ty,
+        dec,
+      );
+      (dec as Dec).ty.constrain(upTy);
+      metadata.define(dec as Dec);
+    } else {
+      throw new Error(`Can't get declaration value from most recent node: ${dec}`);
+    }
+    stmts.push(dec);
+    return stmts;
+  }
+
+  static generate(stmt: Stmt) {
+    return new Dec(
+      stmt.ast,
+      false,
+      genName(),
+      null,
+      stmt,
+    );
+  }
+
+  ref(): Ref {
+    return new Ref(this.ast, this);
+  }
+
+  exprTy(): Type {
+    throw new Error(`declarations can't be used as expressions`);
+  }
+
+  inline(amm: Output) {
+    const name = this.ammName;
+    let ty: Builtin;
+    try {
+      ty = this.ty.breakdown();
+    } catch (e) {
+      console.log('~~~', this);
+      throw e;
+    }
+    if (this.val instanceof Call) {
+      const kind = this.mutable ? 'let' : 'const';
+      this.val.inline(amm, name, ty, kind);
+    } else if (this.val instanceof Lit) {
+      // don't copy the global value, just use it whenever this declaration is used
+      this.__ammName = amm.global('const', this.val.ty.breakdown(), this.val.val);
+    } else {
+      throw new Error(`unexpected expression: ${this.val}`);
+    }
+  }
+}
+
+export class FnArg extends Dec {
+  get ammName(): string {
+    if (super.val != null) {
+      if (!(super.val instanceof Ref)) {
+        throw new Error(`expected fn arg to be set to a reference (here it is: ${super.val})`);
+      }
+      return super.val.ammName;
+    } else {
+      return this.name;
+    }
+  }
+
+  set ammName(to: string) {
+    if (super.val === null) {
+      throw new Error(`cannot rename variables that aren't being inlined`);
+    } else {
+      super.ammName = to;
+    }
+  }
+
+  constructor(
+    ast: LPNode,
+    name: string,
+    ty: Type,
+  ) {
+    super(ast, true, name, ty, null);
+    if (ty === null) {
+      TODO('function params without a type specified');
+    }
+  }
+
+  static fromArgAst(ast: LPNode, metadata: MetaData): FnArg {
+    let name = ast.get('variable').t;
+    let typename = ast.get('fulltypename');
+    let argTy = Type.getFromTypename(typename, metadata.scope);
+    if (argTy === null) {
+      TODO('args with implicit types');
+    } else if (!(argTy instanceof Type)) {
+      throw new Error(`Function argument is not a valid type: ${typename.t}`);
+    }
+    const arg = new FnArg(ast, name, argTy);
+    metadata.define(arg);
+    const dec = metadata.var(arg.name);
+    if (dec !== arg) {
+      throw new Error('ugggghhhhh');
+    }
+    return arg;
+  }
+
+  ammOut(): [string, Builtin] {
+    return TODO('TODO:')
```

Comment:
Better TODO here?

Reply:
that's for one of my extensions that highlights `TODO:` but not `TODO` - this shouldn't get called at all (it's from a previous abstract fn that I defined but didn't get rid of) - will delete